File: bindings/python/protocoll/transport.py
"""Transport abstraction for network communication."""

from protocoll._ffi import ffi, lib
from protocoll.state import _check_error


class Transport:
    """Network transport handle (owns the underlying C resource)."""

    def __init__(self, raw):
        self._raw = raw
        self._closed = False

    @classmethod
    def loopback(cls, bus_id: int = 0):
        """Create a loopback transport for testing.

        Transports with the same bus_id can communicate with each other.
        """
        raw = lib.pcol_transport_loopback_create(bus_id)
        if raw == ffi.NULL:
            raise RuntimeError("Failed to create loopback transport")
        return cls(raw)

    @classmethod
    def udp(cls):
        """Create a UDP transport."""
        raw = lib.pcol_transport_udp_create()
        if raw == ffi.NULL:
            raise RuntimeError("Failed to create UDP transport")
        return cls(raw)

    def bind(self, address: str, port: int):
        """Bind to a local address and port."""
        ep = ffi.new("PcolEndpoint *")
        self._addr_buf = ffi.new("char[]", address.encode())
        ep.address = self._addr_buf
        ep.port = port
        _check_error(lib.pcol_transport_bind(self._raw, ep[0]))

    def close(self):
        """Close and destroy the transport."""
        if not self._closed and self._raw != ffi.NULL:
            lib.pcol_transport_destroy(self._raw)
            self._closed = True

    def __del__(self):
        # Don't destroy if owned by a Peer
        pass

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

File: bindings/rust/protocoll/src/transport.rs
//! Transport abstraction for network communication.

use crate::error::{self, Result};
use std::ffi::CString;

/// Network transport handle (owns the underlying C resource).
pub struct Transport {
    raw: *mut protocoll_sys::PcolTransport,
}

// Safety: The underlying C API uses single-thread ownership. The Peer
// ensures exclusive access. Send+Sync are needed for Arc<Peer> in tests.
unsafe impl Send for Transport {}
unsafe impl Sync for Transport {}

impl Transport {
    /// Create a UDP transport.
    pub fn udp() -> Result<Self> {
        let raw = unsafe { protocoll_sys::pcol_transport_udp_create() };
        if raw.is_null() {
            return Err(crate::Error::Internal);
        }
        Ok(Self { raw })
    }

    /// Create a loopback transport for testing.
    /// Transports with the same `bus_id` can communicate with each other.
    pub fn loopback(bus_id: u32) -> Result<Self> {
        let raw = unsafe { protocoll_sys::pcol_transport_loopback_create(bus_id) };
        if raw.is_null() {
            return Err(crate::Error::Internal);
        }
        Ok(Self { raw })
    }

    /// Bind to a local address and port.
    pub fn bind(&self, address: &str, port: u16) -> Result<()> {
        let addr_c = CString::new(address).map_err(|_| crate::Error::Invalid)?;
        let ep = protocoll_sys::PcolEndpoint {
            address: addr_c.as_ptr(),
            port,
        };
        let code = unsafe { protocoll_sys::pcol_transport_bind(self.raw, ep) };
        error::check(code)
    }

    /// Get the raw pointer (for passing to Peer).
    pub(crate) fn as_raw(&self) -> *mut protocoll_sys::PcolTransport {
        self.raw
    }

    /// Consume self and return the raw pointer without destroying it.
    /// Used when transferring ownership to a Peer.
    pub(crate) fn into_raw(self) -> *mut protocoll_sys::PcolTransport {
        let raw = self.raw;
        std::mem::forget(self);
        raw
    }
}

impl Drop for Transport {
    fn drop(&mut self) {
        if !self.raw.is_null() {
            unsafe { protocoll_sys::pcol_transport_destroy(self.raw) };
        }
    }
}
